sql/rls: push join through permeable Barriers when ON filters are leakproof

spilchen opened the pull request:
Adds normalization rules to push InnerJoin and InnerJoinApply operators below a permeable Barrier when all ON clause filters are leakproof. The Barrier is rewrapped above the join to preserve its semantics while allowing better plan shapes.

This change helps optimize queries affected by RLS, which typically involve barriers around injected RLS predicates.

Note: A similar rule for semi-joins was considered, but no test scenario was found where a barrier incorrectly appears below a semi-join. Existing optgen rules seem to avoid that situation. The semi-join query remains in the join test I added.

Closes #146952

Epic: CRDB-48807
Release note: none

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148531)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner)

___
*[`pkg/sql/opt/norm/rules/join.opt` line 860 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/148531#-OTU0Xeo00nKD4cYqSvs:-OTU0Xep6fYOkuDub74D:bfva9l) ([raw file](https://github.com/cockroachdb/cockroach/blob/c306137a02d6df89752b661b58c2de1e6feb0abf/pkg/sql/opt/norm/rules/join.opt#L860)):*
> ```
>         $leakproofPermeable:* & (If $leakproofPermeable)
>     )
>     $right:*
> ```

I can't think of a case that it would matter in practice, but is this missing the possibility of non-leakproof operations in the other input of the join?


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148531)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @mgartner)

___
*[`pkg/sql/opt/norm/rules/join.opt` line 860 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/148531#-OTU0Xeo00nKD4cYqSvs:-OTXin7XEj57Ct1jvhsE:b-btqw3m) ([raw file](https://github.com/cockroachdb/cockroach/blob/c306137a02d6df89752b661b58c2de1e6feb0abf/pkg/sql/opt/norm/rules/join.opt#L860)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

I can't think of a case that it would matter in practice, but is this missing the possibility of non-leakproof operations in the other input of the join?
</blockquote></details>

Thanks for taking a look. 

Sorry, I'm not sure if I follow. It's only going to promote a barrier that was marked as permeable, which for now is only a barrier added for a RLS filter. So, if there is something on the right side that has a non-leakproof operation, it will either (a) stay protected by a barrier if one exists, or (b) shouldn't matter when it comes to leaking information because it wasn't protected with a barrier to begin with. I'm sure, I'm missing something though. Maybe can you describe it some other way?


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148531)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @spilchen)

___
*[`pkg/sql/opt/norm/rules/join.opt` line 860 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/148531#-OTU0Xeo00nKD4cYqSvs:-OTXornw5AZ7waebzOnX:bqqnltl) ([raw file](https://github.com/cockroachdb/cockroach/blob/c306137a02d6df89752b661b58c2de1e6feb0abf/pkg/sql/opt/norm/rules/join.opt#L860)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Thanks for taking a look. 

Sorry, I'm not sure if I follow. It's only going to promote a barrier that was marked as permeable, which for now is only a barrier added for a RLS filter. So, if there is something on the right side that has a non-leakproof operation, it will either (a) stay protected by a barrier if one exists, or (b) shouldn't matter when it comes to leaking information because it wasn't protected with a barrier to begin with. I'm sure, I'm missing something though. Maybe can you describe it some other way?
</blockquote></details>

For example, say we have a scenario where the barrier is on the left input, and the right input has a non-leakproof filter on the same column as the one used in the join's condition:
```
(Join
  (Barrier (Select (Scan xy) (RLS filters)))
  (Select (Scan ab) (a/0 = 0))
  (x = a)
)
```
You could imagine a rule that infers that the non-leakproof filter applies to both sides of the join because of the equality. So, after the barrier moves above the join, that hypothetical rule results in a situation like this one:
```
(Barrier
  (Join
    (Select (Scan xy) (x/0 = 0) (RLS filters))
    (Select (Scan ab) (a/0 = 0))
    (x = a)
  )
)
```
It's difficult to say whether something like this could happen with the current set of optimizer rules, but we need to make sure it's not possible at all, and also protect against future rules.


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148531)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @mgartner)

___
*[`pkg/sql/opt/norm/rules/join.opt` line 860 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/148531#-OTU0Xeo00nKD4cYqSvs:-OTbJyE2Ci6F-0i9ridD:b7e479d) ([raw file](https://github.com/cockroachdb/cockroach/blob/c306137a02d6df89752b661b58c2de1e6feb0abf/pkg/sql/opt/norm/rules/join.opt#L860)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

For example, say we have a scenario where the barrier is on the left input, and the right input has a non-leakproof filter on the same column as the one used in the join's condition:
```
(Join
  (Barrier (Select (Scan xy) (RLS filters)))
  (Select (Scan ab) (a/0 = 0))
  (x = a)
)
```
You could imagine a rule that infers that the non-leakproof filter applies to both sides of the join because of the equality. So, after the barrier moves above the join, that hypothetical rule results in a situation like this one:
```
(Barrier
  (Join
    (Select (Scan xy) (x/0 = 0) (RLS filters))
    (Select (Scan ab) (a/0 = 0))
    (x = a)
  )
)
```
It's difficult to say whether something like this could happen with the current set of optimizer rules, but we need to make sure it's not possible at all, and also protect against future rules.
</blockquote></details>

Okay, I think I understand what you're saying now.

To address this, I was thinking of introducing a helper function in the join rule to capture the safety concern you're describing. Here's the idea:

```
func extractFilters(expr memo.RelExpr) (memo.FiltersExpr, bool) {
	// If it's a plain Select, return its filters.
	if sel, ok := expr.(*memo.SelectExpr); ok {
		return sel.Filters, true
	}

	// If it's a Barrier, look inside the protected input.
	if barrier, ok := expr.(*memo.BarrierExpr); ok {
		return extractFilters(barrier.Input)
	}

	// Not a recognized structure that holds filters.
	return nil, false
}

func (c *CustomFuncs) CanLiftBarrierAboveJoin(
	onFilters memo.FiltersExpr, barrierSide memo.RelExpr, otherSide memo.RelExpr,
) bool {
	// Early out if any of the ON filters are not leakproof.
	for i := range onFilters {
		if !onFilters[i].ScalarProps().VolatilitySet.IsLeakproof() {
			return false
		}
	}

	barrierCols := barrierSide.Relational().OutputCols
	filters, ok := extractFilters(otherSide)
	if !ok {
		return false // can't inspect filters -> conservatively block the rule
	}

	for _, filter := range filters {
		if !filter.ScalarProps().VolatilitySet.IsLeakproof() {
			filterCols := filter.ScalarProps().OuterCols

			for _, joinFilter := range onFilters {
				cond := joinFilter.Condition
				eq, ok := cond.(*memo.EqExpr)
				if !ok {
					return false // cannot verify safety, be conservative
				}

				left, ok1 := eq.Left.(memo.ScalarPropsExpr)
				right, ok2 := eq.Right.(memo.ScalarPropsExpr)
				if !ok1 || !ok2 {
					return false // cannot verify safety, be conservative
				}

				leftCols := left.ScalarProps().OuterCols
				rightCols := right.ScalarProps().OuterCols

				if (barrierCols.Intersects(leftCols) && filterCols.Intersects(rightCols)) ||
					(barrierCols.Intersects(rightCols) && filterCols.Intersects(leftCols)) {
					return false
				}
			}
		}
	}
	// All filters are leakproof or provably isolated from protected columns.
	return true
}
```

A couple of things I'm not sure of:
-   right now, I only check for `EqExpr` join conditions (i.e., simple equalities like `x = a`). Should this be made more general, or do you think this is correct to just check for equality now?
-   as a safety measure, if I can't analyze the join conditions (e.g., due to unknown or non-equality expressions), I conservatively return `false` and block the barrier from being lifted. Do you think this is the correct way to go?
-  since the kind of plan you described can't be constructed via SQL today, I'm trying to write a test in `exprnorm` that directly builds the expression. But I'm running into issues with the `Barrier` expression. It has a `bool` field, which doesn't map cleanly in `exprnorm`, where all components need to correspond to memo structs. To unblock this, I'm considering changing the `Barrier` to take a `ScalarExpr` instead of a raw `bool`, and using `TrueSingleton`/`FalseSingleton` internally to avoid unnecessary allocations. Does that direction make sense to you?


<!-- Sent from Reviewable.io -->

mgartner commented:
*[`pkg/sql/opt/norm/rules/join.opt` line 860 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/148531#-OTU0Xeo00nKD4cYqSvs:-OTcXKtT8Rx5Xr-LBjSN:b-8lquos) ([raw file](https://github.com/cockroachdb/cockroach/blob/c306137a02d6df89752b661b58c2de1e6feb0abf/pkg/sql/opt/norm/rules/join.opt#L860)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Okay, I think I understand what you're saying now.

To address this, I was thinking of introducing a helper function in the join rule to capture the safety concern you're describing. Here's the idea:

```
func extractFilters(expr memo.RelExpr) (memo.FiltersExpr, bool) {
	// If it's a plain Select, return its filters.
	if sel, ok := expr.(*memo.SelectExpr); ok {
		return sel.Filters, true
	}

	// If it's a Barrier, look inside the protected input.
	if barrier, ok := expr.(*memo.BarrierExpr); ok {
		return extractFilters(barrier.Input)
	}

	// Not a recognized structure that holds filters.
	return nil, false
}

func (c *CustomFuncs) CanLiftBarrierAboveJoin(
	onFilters memo.FiltersExpr, barrierSide memo.RelExpr, otherSide memo.RelExpr,
) bool {
	// Early out if any of the ON filters are not leakproof.
	for i := range onFilters {
		if !onFilters[i].ScalarProps().VolatilitySet.IsLeakproof() {
			return false
		}
	}

	barrierCols := barrierSide.Relational().OutputCols
	filters, ok := extractFilters(otherSide)
	if !ok {
		return false // can't inspect filters -> conservatively block the rule
	}

	for _, filter := range filters {
		if !filter.ScalarProps().VolatilitySet.IsLeakproof() {
			filterCols := filter.ScalarProps().OuterCols

			for _, joinFilter := range onFilters {
				cond := joinFilter.Condition
				eq, ok := cond.(*memo.EqExpr)
				if !ok {
					return false // cannot verify safety, be conservative
				}

				left, ok1 := eq.Left.(memo.ScalarPropsExpr)
				right, ok2 := eq.Right.(memo.ScalarPropsExpr)
				if !ok1 || !ok2 {
					return false // cannot verify safety, be conservative
				}

				leftCols := left.ScalarProps().OuterCols
				rightCols := right.ScalarProps().OuterCols

				if (barrierCols.Intersects(leftCols) && filterCols.Intersects(rightCols)) ||
					(barrierCols.Intersects(rightCols) && filterCols.Intersects(leftCols)) {
					return false
				}
			}
		}
	}
	// All filters are leakproof or provably isolated from protected columns.
	return true
}
```

A couple of things I'm not sure of:
-   right now, I only check for `EqExpr` join conditions (i.e., simple equalities like `x = a`). Should this be made more general, or do you think this is correct to just check for equality now?
-   as a safety measure, if I can't analyze the join conditions (e.g., due to unknown or non-equality expressions), I conservatively return `false` and block the barrier from being lifted. Do you think this is the correct way to go?
-  since the kind of plan you described can't be constructed via SQL today, I'm trying to write a test in `exprnorm` that directly builds the expression. But I'm running into issues with the `Barrier` expression. It has a `bool` field, which doesn't map cleanly in `exprnorm`, where all components need to correspond to memo structs. To unblock this, I'm considering changing the `Barrier` to take a `ScalarExpr` instead of a raw `bool`, and using `TrueSingleton`/`FalseSingleton` internally to avoid unnecessary allocations. Does that direction make sense to you?
</blockquote></details>

Great catch @DrewKimball! This rule is effectively pushing both the join filters and the entire RHS into the barrier. Therefore, I think the RHS needs to be leakproof in order for this to be valid in the general case. So I think we can fix this by adding this custom function [here](https://github.com/cockroachdb/cockroach/blob/bfbc0b61f2810249088b84d57ee77b1a2a0574b3/pkg/sql/opt/norm/general_funcs.go#L637) and calling it on `$right` in the pattern:

```go
func (c *CustomFuncs) IsLeakproof(expr memo.RelExpr) bool {
	return expr.Relational().VolatilitySet.IsLeakproof()
}
```

This will prevent some possible optimizations for specific RHS expressions. We can write more specialized rules for those cases as they arrive.


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148531)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @mgartner)

___
*[`pkg/sql/opt/norm/rules/join.opt` line 860 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/148531#-OTU0Xeo00nKD4cYqSvs:-OThLDGc24ZAkKm7--LI:buq583j) ([raw file](https://github.com/cockroachdb/cockroach/blob/c306137a02d6df89752b661b58c2de1e6feb0abf/pkg/sql/opt/norm/rules/join.opt#L860)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Great catch @DrewKimball! This rule is effectively pushing both the join filters and the entire RHS into the barrier. Therefore, I think the RHS needs to be leakproof in order for this to be valid in the general case. So I think we can fix this by adding this custom function [here](https://github.com/cockroachdb/cockroach/blob/bfbc0b61f2810249088b84d57ee77b1a2a0574b3/pkg/sql/opt/norm/general_funcs.go#L637) and calling it on `$right` in the pattern:

```go
func (c *CustomFuncs) IsLeakproof(expr memo.RelExpr) bool {
	return expr.Relational().VolatilitySet.IsLeakproof()
}
```

This will prevent some possible optimizations for specific RHS expressions. We can write more specialized rules for those cases as they arrive.
</blockquote></details>

Wow, that's much easier than what I was trying to do. I have applied this change.


<!-- Sent from Reviewable.io -->

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Just a few minor suggestions. @DrewKimball this change still LGTY, correct?

> Reviewed 1 of 5 files at r1, 4 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148531#-:-OTmqPBo-PMRgiXWFLsO:b-wsrf49)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball)

___
*[`pkg/sql/opt/norm/testdata/rules/join` line 64 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/148531#-OTmoMdHCMI8JY2YbWYc:-OTmoMdHCMI8JY2YbWYd:b-2m5gb0) ([raw file](https://github.com/cockroachdb/cockroach/blob/3721d8e776f5781070a9c03a8eadf2ea9360dbd4/pkg/sql/opt/norm/testdata/rules/join#L64)):*
> ```
> LANGUAGE SQL IMMUTABLE LEAKPROOF
> AS $$
>   SELECT count(*) FROM system.users; -- to prevent the function from being inlined
> ```

Another option that let's you exercise the same test cases without added the extra statement is to disable the rule that inlines UDFs for those tests:

```
opt disable=InlineUDF expect=...
SELECT ...
----
```
___
*[`pkg/sql/opt/norm/testdata/rules/join` line 4570 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/148531#-OTmmVF2BZOrpoL7m0FP:-OTmmVF2BZOrpoL7m0FQ:bmqcamf) ([raw file](https://github.com/cockroachdb/cockroach/blob/3721d8e776f5781070a9c03a8eadf2ea9360dbd4/pkg/sql/opt/norm/testdata/rules/join#L4570)):*
> ```
> 
> # --------------------------------------------------
> # PushLeakproofJoinLeftIntoPermeableBarrier / PushLeakproofJoinRightIntoPermeableBarrier
> ```

nit: I'd just split this into two lines
___
*[`pkg/sql/opt/norm/rules/join.opt` line 849 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/148531#-OTmn_ZODcZLaLxqNydx:-OTmn_ZODcZLaLxqNydy:b-hlwa1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/3721d8e776f5781070a9c03a8eadf2ea9360dbd4/pkg/sql/opt/norm/rules/join.opt#L849)):*
> ```
> )
> 
> # PushLeakproofJoinLeftIntoPermeableBarrier moves a join below a permeable
> ```

nit: Maybe more accurate names would be "PushLeakproofJoinIntoPermeableBarrierLeft"/"PushLeakproofJoinIntoPermeableBarrierRight". Or "PullPermeableBarrierAboveJoin"?
___
*[`pkg/sql/opt/norm/rules/join.opt` line 853 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/148531#-OTmmBTM0U_QG34yaGEB:-OTmmBTM0U_QG34yaGEC:b-pq0xt1) ([raw file](https://github.com/cockroachdb/cockroach/blob/3721d8e776f5781070a9c03a8eadf2ea9360dbd4/pkg/sql/opt/norm/rules/join.opt#L853)):*
> ```
> # then placed above the join. This is safe because leakproof filters can be
> # reordered freely, and the Barrier allows such movement when marked as
> # LeakproofPermeable.
> ```

nit: Add a sentence like: "This rule is effectively pushing the RHS into the barrier, so the RHS must be leakproof as well."
___
*[`pkg/sql/opt/norm/rules/join.opt` line 873 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/148531#-OTmmLJJAbziDh7YdsKf:-OTmmLJJAbziDh7YdsKg:bhglcr) ([raw file](https://github.com/cockroachdb/cockroach/blob/3721d8e776f5781070a9c03a8eadf2ea9360dbd4/pkg/sql/opt/norm/rules/join.opt#L873)):*
> ```
> # It moves a join below a permeable Barrier on its right input when all ON
> # filters are leakproof, then rewraps the join in the Barrier to preserve its
> # blocking behavior for non-leakproof expressions higher in the plan.
> ```

nit: Add a sentence like: "This rule is effectively pushing the LHS into the barrier, so the LHS must be leakproof as well."


<!-- Sent from Reviewable.io -->

spilchen committed:
sql/rls: push join through permeable Barriers when ON filters are leakproof

Adds normalization rules to push InnerJoin and InnerJoinApply operators
below a permeable Barrier when all ON clause filters are leakproof. The
Barrier is rewrapped above the join to preserve its semantics while
allowing better plan shapes.

This change helps optimize queries affected by RLS, which typically
involve barriers around injected RLS predicates.

Note: A similar rule for semi-joins was considered, but no test scenario
was found where a barrier incorrectly appears below a semi-join. Existing
optgen rules seem to avoid that situation. The semi-join query remains
in the join test I added.

Closes #146952

Epic: CRDB-48807
Release note: none

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148531)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball and @mgartner)

___
*[`pkg/sql/opt/norm/rules/join.opt` line 849 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/148531#-OTmn_ZODcZLaLxqNydx:-OU-bqfDAKiQTaUj8I_e:b-578zzp) ([raw file](https://github.com/cockroachdb/cockroach/blob/3721d8e776f5781070a9c03a8eadf2ea9360dbd4/pkg/sql/opt/norm/rules/join.opt#L849)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: Maybe more accurate names would be "PushLeakproofJoinIntoPermeableBarrierLeft"/"PushLeakproofJoinIntoPermeableBarrierRight". Or "PullPermeableBarrierAboveJoin"?
</blockquote></details>

I went with PushLeakproofJoinIntoPermeableBarrierLeft/Right.


<!-- Sent from Reviewable.io -->
